### PR TITLE
Add command expansions %key{body}

### PR DESCRIPTION
**Pull request by ksdrar**
Based on: https://github.com/helix-editor/helix/pull/3393

### CODE REVIEW

**Review by the-mikedavis — CHANGES_REQUESTED** [review]
For this PR we only touch typable commands but we will probably want to re-use expansion elsewhere (for example custom commands (https://github.com/helix-editor/helix/pull/3393#issuecomment-1506091918) or exposing the file-type in formatter arguments (https://github.com/helix-editor/helix/pull/5626)). So I think we can formalize expansion a little by moving the replacement function into the editor module in helix-view. I'd also like to see a few unit tests of the regex.

We could also add some integration tests if we add an echo command:

```rust
// helix-term/src/commands/typed.rs
fn echo(cx: &mut compositor::Context, args: &[Cow<str>], event: PromptEvent) -> anyhow::Result<()> {
    if event != PromptEvent::Validate {
        return Ok(());
    }

    cx.editor.set_status(format!("{}", args.join(" ")));
    Ok(())
}
```

(for example could be used like `:echo %val{filename}` or `:echo %sh{echo %{filename}}`)

File: helix-term/src/commands.rs
Comment (ksdrar) [reply helix-term/src/commands.rs]: I think I got what you mean, can you check the latest commit please

File: helix-term/src/commands/typed.rs
Comment (ksdrar) [reply helix-term/src/commands/typed.rs:2920]: I think dirname migth be confused with cwd by users

File: helix-term/src/commands/typed.rs
Comment (ksdrar) [reply helix-term/src/commands/typed.rs:2907]: I change it to allow sh body to expand using recursion, but let me check if it can be done

File: helix-term/src/commands/typed.rs
Comment (ksdrar) [reply helix-term/src/commands/typed.rs:2920]: Oh, didn't know that. In that case it's fine to change it

**Comment by ksdrar** [comment]
%sh{body} uses `shell_impl` from helix-term/src/commands.rs which cannot be accessed from helix-view, what should be done in this case?

File: helix-term/src/commands/typed.rs
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:2202]: This doesn't need to be a function on its own, it can be moved back into the prompt definition in `command_mode`

File: helix-view/src/editor.rs
Comment (the-mikedavis) [inline helix-view/src/editor.rs:692]: This was recently addressed in #7013 so if you rebase then this line should disappear from the diff

File: helix-view/src/editor.rs
Comment (the-mikedavis) [inline helix-view/src/editor.rs:1784]: I don't think we need to log here since we're returning and handling the Err variant. `Editor::set_error` already logs each error message

File: helix-view/src/editor.rs
Comment (the-mikedavis) [inline helix-view/src/editor.rs:1729]:
```suggestion
        &EXPAND_ARGS_REGEXP,
```

Comment (the-mikedavis) [reply helix-view/src/editor.rs:1729]: Also, a minor nit that this constant should probably be called `REGEX` rather than `REGEXP` to be consistent with regex in the regex crate and other usages in Helix

File: helix-view/src/editor.rs
Comment (the-mikedavis) [inline helix-view/src/editor.rs:1724]:
```suggestion
pub fn expand_variables<'a>(editor: &Editor, input: &'a str) -> anyhow::Result<Cow<'a, str>> {
```

This will eventually be used for things other than arguments

File: helix-view/src/editor.rs
Comment (ksdrar) [reply helix-view/src/editor.rs:692]: I don't think it was changed, I'll save the file without formatting that line if you want

File: helix-term/src/commands/typed.rs
Comment (ksdrar) [inline helix-term/src/commands/typed.rs:2883]: Should it print the error and return if `expand_variables` fails?

**Comment by pascalkuthe** [comment]
> So I think we can formalize expansion a little by moving the replacement function into the editor module in helix-view. 

I personally would prefer this as a separate module so `helix-view/src/editor/substitute.rs` I think `editors.rs` is already a bit too much of  a kitchen sink and this is pretty orthogonal so it makes more sense as its own module IMO

**Comment by ksdrar** [comment]
Like this?

```rust
// helix-view/src/editor.rs
mod variable_expansion;
pub use variable_expansion::expand_variables;
...
```

```rust
// helix-view/src/editor/variable_expansion.rs
pub fn expand_variables<'a>(...
```

**Comment by pascalkuthe** [comment]
yeah exactly :+1:

**Review by pascalkuthe — APPROVED** [review]
LGTM :+1:  Just missing some tests

**Comment by jlebon** [comment]
Awesome, looking forward to this! Can we document this somewhere, esp. the exact variable names supported?

Also, should this be marked as closing https://github.com/helix-editor/helix/issues/3134?

**Comment by ksdrar** [comment]
Checking out https://github.com/helix-editor/helix/pull/7090 I think it's a good idea for users to be able to expand the variables while in `:command` mode and maybe even add the variables to the list of suggestions for autocompletion. What do you think?

**Comment by ksdrar** [comment]
I just pushed this feature into a new branch of my fork (ce-prompt).

Input inside a Prompt (command_mode, file_picker, etc) can be expanded from variables to values using Ctrl + .

https://github.com/helix-editor/helix/assets/22417151/25befd25-8e7f-4195-b458-3f7b9c4d6c6a

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:207]:
```suggestion
                        .map(|arg| helix_view::editor::expand_variables(cx.editor, arg))
```

**Review by the-mikedavis — COMMENTED** [review]
I think this is looking good. @archseer had the idea of replacing the `%val{varname}` syntax with just `%varname`, for example `%filename`, `%dirname`, etc.. `%sh{..}` would stay the same. I think Kakoune's syntax makes more sense for Kakoune which mostly integrates with scripting / shell. I don't forsee us needing other `%X{..}`s other than maybe one for the plugin system language (though maybe this whole syntax/feature would change with the plugin system). What do you think?

**Comment by danillos** [comment]
@ksdrar Today I updated to use this new branch and this command is not working any more for me.

```
s = ':open %sh{~/.dotfiles/support/pipe.rb switch_to_spec %sh{pwd} %val{filename}}'
```

It opens 4 buffers:

1. `%sh{~/.dotfiles/support/pipe.rb`
2. `switch_to_spec`
3. `~/my-project-pwd`
4. `}`

**Comment by ksdrar** [comment]
> I think this is looking good. @archseer had the idea of replacing the `%val{varname}` syntax with just `%varname`, for example `%filename`, `%dirname`, etc.. `%sh{..}` would stay the same. I think Kakoune's syntax makes more sense for Kakoune which mostly integrates with scripting / shell. I don't forsee us needing other `%X{..}`s other than maybe one for the plugin system language (though maybe this whole syntax/feature would change with the plugin system). What do you think?

I think it's better to leave it be `%val{...}` to avoid issues with something like `%filename.bak` or unintended replacements. Or maybe changing it to `%{varname}` to avoid `val` repetition.

**Comment by David-Else** [comment]
@ksdrar Hope you are well. Is this nearly ready to go? I think it just misses some docs and implementing the-mikedavis's suggested change?

**Comment by ksdrar** [comment]
Hello @David-Else, I just pushed a commit to "cmd-exp-2" branch (it's rebased to latest changes) in my fork changing the expansion method and also uses the proposed %sh{} and %{varname} changes. Can you or someone else try it, please.

**Comment by David-Else** [comment]
@ksdrar I tried the latest pull request here with the `:sh echo %val{filename}` format and it works great. I use this script and the `gh` cli command:

```sh
ghpr() { GH_FORCE_TTY=100% gh pr list --limit 300 |
    fzf --ansi --preview 'GH_FORCE_TTY=100% gh pr view {1}' --preview-window 'down,70%' --header-lines 3 |
    awk '{print $1}' |
    xargs gh pr checkout; }
```

I am afraid I don't know how to pull in your branch `https://github.com/ksdrar/helix/tree/cmd-exp-2` using git commands so it creates a new branch in my helix source code repository on my local machine. If you could tell me I would be happy to try it out... or just push the changes here and I will test them :)

**Comment by danillos** [comment]
@ksdrar I'll try it today. I'm a heavy user of this feature. But I'm using the previous branch, this one had some bugs the last time I tested.

**Comment by ksdrar** [comment]
> @ksdrar I'll try it today. I'm a heavy user of this feature. But I'm using the previous branch, this one had some bugs the last time I tested, reported [here](https://github.com/helix-editor/helix/pull/6979#issuecomment-1603370943)

Thanks!, I tried your command and it works fine (I just tried parsing it as I obviously don't have your ruby script). But if you have any problem, report back please :)

**Comment by ksdrar** [comment]
> @ksdrar I tried the latest pull request here with the `:sh echo %val{filename}` format and it works great. I use this script and the `gh` cli command to audition pull requests:
> 
> ```shell
> ghpr() { GH_FORCE_TTY=100% gh pr list --limit 300 |
>     fzf --ansi --preview 'GH_FORCE_TTY=100% gh pr view {1}' --preview-window 'down,70%' --header-lines 3 |
>     awk '{print $1}' |
>     xargs gh pr checkout; }
> ```
> 
> I am afraid I don't know how to pull in your branch `https://github.com/ksdrar/helix/tree/cmd-exp-2` using git commands so it creates a new branch in my helix source code repository on my local machine. If you could tell me I would be happy to try it out... or just push the changes here and I will test them :)

You may try the branch using the following commands.
`git remote add ksdrar https://github.com/ksdrar/helix`
`git pull ksdrar cmd-exp-2`
`git switch cmd-exp-2`

Or if you don't want to create a branch in your local machine (not recommended).
`git remote add ksdrar https://github.com/ksdrar/helix`
`git merge ksdrar/cmd-exp-2`

**Comment by David-Else** [comment]
@ksdrar I followed your instructions and it works great! For example, this works perfectly:

```
:sh echo %{filename}:%{linenumber}
```
@danillos I think you are better qualified to test the expansions are all correct as per your post https://github.com/helix-editor/helix/pull/3393#issuecomment-1520996787

**Comment by danillos** [comment]
@ksdrar I did some tests and found these issues:

Issue 1:
When executing  `:sh echo %sh{echo a}` it works showing `a`
but the same command in a keybind `t = ':sh echo %sh{echo a}'` it shows `%sh{echo a}`

Issue 2:
`:sh echo %sh{echo %{filename} %sh{pwd}}` it works
`:sh echo %sh{echo %{filename} %sh{pwd} %{filename}}` it doesn't works, it seems that any args after the inner %sh breaks the logic

**Comment by ksdrar** [comment]
@danillos I just pushed a commit for the second issue. I tried and it was caused by not trimming the command output. Can you try it please.

About the first issue, the thing is the keybind arguments are parsed by separating the string and saving it as a vector, so the command arguments end up being ["%sh{echo", "a}"]. We would have to join the vector again and then separate it once more for it to work with the expansion. It was previously spotted by @the-mikedavis in https://github.com/helix-editor/helix/pull/6979#discussion_r1191574086. I think we should join them to be parsed.

**Comment by danillos** [comment]
@ksdrar Confirmed, Issue 2 was fixed.

**Comment by ksdrar** [comment]
> @ksdrar Confirmed, Issue 2 was fixed.

Thanks!. I just pushed a commit addressing the first issue, it should work as expected now.

**Comment by danillos** [comment]
@ksdrar Issue 1 was fixed too.

But I'm not allowed to save changes, when I attempt to `:w`, it shows it:
CleanShot 2023-08-26 at 06 02 52@2x

**Comment by David-Else** [comment]
I have an issue with the latest `cmd-exp-2` version. I can't use my shortcut `Z = { Z = ":wq", Q = ":q!" }`, it just says `quit! takes no arguments`, probably the same thing as https://github.com/helix-editor/helix/pull/6979#issuecomment-1694238823

**Comment by ksdrar** [comment]
Yeah, I just pushed another commit. Now it should work fine

**Comment by David-Else** [comment]
> Yeah, I just pushed another commit. Now it should work fine

Fixed!

**Comment by David-Else** [comment]
Can this be used in `languages.toml`? Something like `formatter = { command = 'prettier', args = [ "%{filename}" ] }
`? I can't really experiment with this until https://github.com/helix-editor/helix/pull/6846 is merged.

**Comment by ksdrar** [comment]
> Can this be used in `languages.toml`? Something like `formatter = { command = 'prettier', args = [ "%{filename}" ] } `? I can't really experiment with this until #6846 is merged.

Not really. There's https://github.com/helix-editor/helix/pull/5626 and it's only being held back because the variable declaration syntax (%val{varname} or %{varname}, etc) is not set in stone yet afaik. I think once this PR is merged the formatter one will follow.

**Comment by danillos** [comment]
@ksdrar, everything is working fine on the branch `cmd-exp-2`.

**Comment by David-Else** [comment]
Currently, `filename` provides the absolute path, while the `dirname` command gives the path up until the filename. Is it possible to obtain just the filename? Wouldn't it be more logical for the `filename` to only provide the file name? We could easily create the absolute path by using `dirname filename`, and get the bonus of having access to just the file name. Just an idea, I am probably missing some context here.

**Comment by quantonganh** [comment]
> Is it possible to obtain just the filename?

You can use [basename](https://linux.die.net/man/1/basename):

```
:sh basename %val{filename}
```

**Comment by ksdrar** [comment]
I just added `basename` and `selection` variables with the latest commit to "cmd-exp-2". But I think renaming `filename` to `filepath` and `basename` to `filename` would make their meaning clearer for some users, what do you think?

**Comment by David-Else** [comment]
> I just added `basename` and `selection` variables with the latest commit to "cmd-exp-2". But I think renaming `filename` to `filepath` and `basename` to `filename` would make their meaning clearer for some users, what do you think?

I agree.

**Comment by David-Else** [comment]
@ksdrar I saw you requested a review, but I think you forgot to push the new `cmd-exp-2` here.

**Comment by ksdrar** [comment]
Yeah, sorry about that. Now the changes are merged with this branch and I added integration tests for variables

**Comment by jiehuifang** [comment]
Helix will merge this pull request in version 23.09 ?

**Comment by MDeiml** [comment]
@ksdrar what's the state of this PR? Just waiting for review?

**Comment by rcorre** [comment]
Looks like the new test is failing on mac: 
https://github.com/helix-editor/helix/actions/runs/6103715977/job/16568045931?pr=6979#step:7:768

**Comment by zaquestion** [comment]
@the-mikedavis is this in a rereviewable state or one where the unittests can be run?

**Comment by the-mikedavis** [comment]
Looks like one of the integration tests fails on windows and the docs need to be updated with `cargo xtask docgen` to add the new command. I'd like to give this another look soon though

**Comment by zaquestion** [comment]
So the windows error (just one) seems to be related to the behavior of [std::fs::canonicalize](https://doc.rust-lang.org/std/fs/fn.canonicalize.html)  which returns windows paths using [extended path syntax](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file).

```
thread 'test::commands::variable_expansion::test_variable_expansion' panicked at 'assertion failed: `(left == right)`
  left: `"C:\\Users\\runneradmin\\AppData\\Local\\Temp\\.tmpr1kjJM"`,
 right: `"\\\\?\\C:\\Users\\runneradmin\\AppData\\Local\\Temp\\.tmpr1kjJM"`', helix-term\tests\test\commands\variable_expansion.rs:29:9
 ```
 The right in this case being what is we are expecting. We can see that the path we actually returned is otherwise correct just not using this syntax. Poking at the code, it seems like we are using paths returned out of helix's Editor crate, which is probably the path type we want to check against (not extended path syntax)? If so, would patching the test to return strip the extended path prefix be suitable here? Perhaps we want to update Editor to return the extended paths? 
 
 If it's just the former I can probably PR this branch with those changes, I might struggle a bit with the later but willing to try with some guidance.

**Comment by pascalkuthe** [comment]
using `std::fs::cannonicalize` is (almost) always a bug you should be using the canonicalization functions from helix core (so canonicalization is always consistent and correct)

**Comment by YurkoHoshko** [comment]
QQ: does this PR require just a doc update and using Helix Core canonicalization functions in test instead of standard library one? If so - I could take a stab at it (not a Rust dev myself, but would love this PR to get merged :))

**Comment by ksdrar** [comment]
Sorry for the delay, I just pushed the docs update and canonicalize changes. Hope it works now.

**Comment by jhugs** [comment]
@the-mikedavis @pascalkuthe can we get some eyes on his changes? Super hyped for this PR! 🎉

**Comment by rcorre** [comment]
Would there be any way to get the primary selection as a line range, as mentioned [in this comment](https://github.com/helix-editor/helix/issues/3134#issuecomment-1426286074)?

I'd really love to use `gh browse` to link to a block of code. Right now it looks like you can get a single line number, or the selection as text, but not the selection as a line range.

File: helix-term/src/ui/mod.rs
Comment (the-mikedavis) [inline helix-term/src/ui/mod.rs:387]: Would the available variables ever change depending on the state of Editor? I think this can drop the `&Editor` parameter

File: helix-term/tests/test/commands/variable_expansion.rs
Comment (the-mikedavis) [inline helix-term/tests/test/commands/variable_expansion.rs:2]: Integration tests shouldn't be calling functions like this directly. Instead these should use `test_key_sequence` with `:echo` and `Editor::get_status`

File: helix-view/src/editor/variable_expansion.rs
Comment (the-mikedavis) [inline helix-view/src/editor/variable_expansion.rs:147]:
```suggestion
        Some(o) => Ok(Cow::Owned(o)),
        None => Ok(Cow::Borrowed(input)),
```

since `Cow` is already in scope

File: helix-view/src/editor/variable_expansion.rs
Comment (the-mikedavis) [inline helix-view/src/editor/variable_expansion.rs:55]:
```suggestion
                                    "cwd" => helix_std::env::current_working_dir().to_str().unwrap().to_owned(),
```

that function will be available if you rebase/merge-master

File: helix-view/src/editor/variable_expansion.rs
Comment (the-mikedavis) [inline helix-view/src/editor/variable_expansion.rs:15]: Let's make this a member function of Editor:

```rust
impl Editor {
    pub fn expand_variables<'a>(&self, input: &'a str) -> anyhow::Result<Cow<'a, str>> {
        // ...
    }
}
```

File: helix-term/src/commands/typed.rs
Comment (the-mikedavis) [inline helix-term/src/commands/typed.rs:2486]: Looks like this function is no longer used and can be removed

File: book/src/generated/typable-cmd.md
Comment (the-mikedavis) [inline book/src/generated/typable-cmd.md:1]: Aside from these docs we should document the variable expansion somewhere in the book. I think a new section in `book/src/usage.md` would probably be the right spot

**Review by the-mikedavis — COMMENTED** [review]
Thanks for continuing to work on this and sorry it's been a while since my last review. This is overall looking pretty good. I left a few comments.

Also \cc @archseer I'm curious what you think of the `%{variable}` syntax over `%variable`. I don't have strong opinions either way and I see how `%variable` might be awkward if you're concatenating the variable with other strings.

File: helix-view/src/editor/variable_expansion.rs
Comment (cole-h) [reply helix-view/src/editor/variable_expansion.rs:55]: (you probably meant `helix_stdx::env::....`, not `helix_std::env::....`)

**Comment by ksdrar** [comment]
I just pushed the changes, can you review again please?

File: book/src/usage.md
Comment (the-mikedavis) [inline book/src/usage.md:217]:
```suggestion
| `%{linenumber}` | The line number where the primary cursor is positioned. |
| `%{selection}`  | The text selected by the primary cursor. |
```

File: helix-term/src/ui/mod.rs
Comment (the-mikedavis) [reply helix-term/src/ui/mod.rs:387]: It looks like this hasn't been resolved yet

Comment (ksdrar) [reply helix-term/src/ui/mod.rs:387]: You mean remove the `&Editor` parameter from the function? I can't do it without breaking the `pub type Completer = fn(&Editor, &str) -> Vec<Completion>` requirement

Comment (the-mikedavis) [reply helix-term/src/ui/mod.rs:387]: Ah right good catch. Let's leave it as-is :+1:

**Review by the-mikedavis — APPROVED** [review]

**Comment by jhugs** [comment]
@pascalkuthe @archseer can we get another review? ❤️ (I'm sorry I'm super impatient for this one 😆)

File: helix-term/src/commands.rs
Comment (pascalkuthe) [inline helix-term/src/commands.rs:210]: I am pretty sure this is not correct. We allow quoting strings so arguments can contain whitespace and a simple split like this will not work. This seems like entirely the wrong place to handle command 3xpansion. This probably needs to be integrated into the command parser properly so that it can correctly interact with whitespace and escapes(without requiring quotes everywhere) so into the shellwords parser somehow. Altough that should likely not actually handle the expansion since it's also used for auto completion. This PR also doens't seem to handle completions at all (`:open %{dirname}/` will not offer useful completions). It should be possible to to properly implement that altough it may be tricky (%sh should not be executed for completions tough)

**Comment by carterols** [comment]
What’s the status of this? This is the one feature that is keeping me from using helix, as I can’t bind any key maps to run custom commands involving the current file name.

**Comment by jhugs** [comment]
> What’s the status of this? This is the one feature that is keeping me from using helix, as I can’t bind any key maps to run custom commands involving the current file name.

@carterols I'm also eagerly awaiting this, but depending on what you want to do, there is a "work around" in which you can parse the current file name via terminal if the one you're using has the functionality (I'm using kitty, the OP used [wezterm](https://github.com/quantonganh/helix-wezterm/blob/92e3ffc7f389c4508c4090564d7181b5bc855fd8/helix-wezterm.sh#L6C1-L6C9))

**Comment by neduard** [comment]
Hey all!  I've been looking for this feature as well (thank you all that have put it together so far!!) and I'm wondering if what we have atm is a bit overkill?  I'm thinking we already have the `:sh` command which is great and the only thing we are really missing are some variable expansions when running it.

For example, `git blame` can be run as `:sh git blame -L %{linenumber} {%filepath}`.

One of the things I really like about helix is that it values a clean code base and doesn't "try to be everything for everyone".  As such I'm wondering if we can simplify (or at least consider having a simpler solution and avoid some of the edge cases):
* `%filepath` type variables I think are ok, `%filepath.bak` is actually totally valid and should expand to the filepath with `.bak` extension.  This is in-line with what shells do eg ` f=123; echo $f.bak` outputs `123.bak` (although might be slightly more difficult to implement?)
* we can drop the `:echo` and `%sh{}` as we already have `:sh`

I'm also wondering if `filepath` / `dirpath` are more readable than `filename` / `dirname`?  To me `filename` implies the name, not the full path.

What do people think?  I've only been using helix for a few months (and loving it! it has pretty much replaced like 12 years of vim) but this is my first time here so really sorry if I'm talking complete rubbish haha! :see_no_evil:

**Comment by postsolar** [comment]
> we can drop the `:echo` and `%sh{}` as we already have `:sh`

How would I then do something like `:open %sh{ fzf }`, without resorting to temporary buffers?

**Comment by carterols** [comment]
> > we can drop the `:echo` and `%sh{}` as we already have `:sh`
> 
> How would I then do something like `:open %sh{ fzf }`, without resorting to temporary buffers?

I understand this use case, but maybe this PR can focus on getting basic support for things like `%{filepath}` and we can later tackle the `%sh{}` expansions and other, more complex, situations?

As a Helix user, I would much rather have features now that cover most use cases than trying to lump everything together in one PR that takes a long time to get approved. I really could use this feature because at work, I am forced to use Perforce as my SCM, and you have to manually checkout a file to make it writable via a `p4 edit <filepath>`. Without the command expansion, I have to manually type this in every time with the full file path, whereas in neovim, I’ve bound this action to a key binding that automatically expands the current file path that is open.

**Comment by danillos** [comment]
> we can drop the :echo and %sh{} as we already have :sh

I have a case that that uses `%sh{}` and I use it a lot every day, so I would like to it not be removed.

It uses `%sh{}` to get the current dir, and to run a script that will find the unit test from a file, of the file from a unit test, and after it sends to helix open it.
```
s = ':open %sh{~/.dotfiles/helix_helper.rb switch_to_spec %sh{pwd} %{filename}}'
```

And I have many commands like this, using `%sh{pwd}`

```
B = ':sh ~/.dotfiles/helix_helper.rb rspec_file %sh{pwd} %{filename}:%{linenumber}'
```

**Comment by David-Else** [comment]
`:echo %sh{git blame -L %{linenumber} %{filename}}` vs `:sh git blame -L %{linenumber} {%filepath}` demonstrates that there is a danger of things becoming needlessly complex. Having `sh` and `%sh` is a potential redundancy that really needs more explanation in the docs, especially for people new to using shell commands.

The main question is what is delaying this very important feature getting merged? It is not much code, so maybe the implementation is still not meeting the core maintainers vision as it is?

**Comment by rcorre** [comment]
@danillos could you not get `pwd` within `helix_helper.rb`?

**Comment by danillos** [comment]
> @danillos could you not get `pwd` within `helix_helper.rb`?

Yes, it works thanks, I don't know why I didn't it before, now I can change to:

```
s = ':open %sh{~/.dotfiles/helix_helper.rb switch_to_spec %{filename}}'
```

**Comment by postsolar** [comment]
There is also `%{cwd}` for this.

**Comment by lizclipse** [comment]
I think it's worth keeping the `%sh{}` expansion as it allows a lot of flexibility when writing custom commands and integrating with other tools. At the very least, it's a handy 'get-out-of-jail' tool in that it allows you to drop into a full shell when building more complex arguments.

The main blockers on this seem to be the whitespace handling and completions. My view is that the WS handling is fine as is, especially since the last time I used `:set` it didn't handle whitespace properly (which was a week or 2 back a I think) and both can be fixed up later, and the completions can (and probably should) be done in a separate PR so that this one can get in and better tested.

Both points probably warrant TODOs, but I don't think should hold up and (otherwise) ready feature

**Comment by noahfraiture** [comment]
Hope this will be merge soon, it would be a great workaround for many plugin

**Comment by tapyu** [comment]
I am impatiently waiting for this PR be merged :eyes:

**Comment by acaloiaro** [comment]
I know this PR has quite a few watchers and I don't want to create noise, but since it has been approved by @the-mikedavis for nearing six months and discussion has seemingly reached a stopping point, can someone say what it needs to be merged? 

Is the pending review from @archseer what the merge is pending? Or is it actually ready to go?

**Comment by ksdrar** [comment]
Sorry guys, I've been busy with work but I'll try to find some time to finish this PR this week. I'll leave the completions feature for a future PR

**Comment by tdaron** [comment]
Hey @ksdrar,

Would you mind if I (or someone else) try to finish this PR if you don't have time? It's almost ready to land according to the comments, and a lot of people are waiting for it to be completed. I'd like to help move it forward.

Thanks!

**Comment by ksdrar** [comment]
Yeah @tdaron! That would be awesome, thanks for your help!

**Comment by tdaron** [comment]
Yay !
I rebased your work on master and removed completion stuff (as it would be integrated in another PR) on my personal fork. 

What's currently missing on this to be ready?

**Comment by ksdrar** [comment]
@pascalkuthe wants to integrate the expansion not in the execution phase but in parsing, but I think that it would require a simple AST or something like that to make the expansion lazy but keeping all the command parsing logic (white space handling, quotes, etc) as it is now. I don't have the path right now, but the structure ShellWords is the parsing phase iirc.

**Comment by David-Else** [comment]
If https://github.com/helix-editor/helix/pull/11164 is now the official new version of this PR, then this should be closed and https://github.com/helix-editor/helix/pull/11164 added to the `next` category instead.

**Comment by quatquatt** [comment]
> > What’s the status of this? This is the one feature that is keeping me from using helix, as I can’t bind any key maps to run custom commands involving the current file name.
> 
> @carterols I'm also eagerly awaiting this, but depending on what you want to do, there is a "work around" in which you can parse the current file name via terminal if the one you're using has the functionality (I'm using kitty, the OP used [wezterm](https://github.com/quantonganh/helix-wezterm/blob/92e3ffc7f389c4508c4090564d7181b5bc855fd8/helix-wezterm.sh#L6C1-L6C9))

Could you share your relevant Kitty workaround?

**Comment by jhugs** [comment]
@quatquatt sure, both my own versions of the wezterm / kitty implementations are [here](https://github.com/jhugs/dotfiles-and-stuff/blob/main/bin/hx_ide) (the wezterm one is in a branch).

**Pull request closed without merging**
